src/mapWithFeedback.test.ts

```diff
@@ -0,0 +1,209 @@
+      const result = mapWithFeedback(
+        [1, 2, 3, 4, 5] as const,
+        (acc, x) => acc + x,
+        100,
+      );
+      expect(result).toEqual([101, 103, 106, 110, 115]);
+    });
+
+    it("should use the same accumulator on every iteration if it's mutable, therefore returning an array containing {array length} references to the accumulator.", () => {
+      const results = mapWithFeedback(
+        [1, 2, 3, 4, 5] as const,
+        (acc, x) => {
+          acc[x] = x;
+          return acc;
+        },
+        {} as Record<string, unknown>,
+      );
+
+      const expectedEquality = Array.from({ length: 5 }).fill({
+        "1": 1,
+        "2": 2,
+        "3": 3,
+        "4": 4,
+        "5": 5,
+      });
+      expect(results).toEqual(expectedEquality);
+      const [item] = results;
+      expect(results.every((result) => result === item)).toBe(true);
+    });
+
+    it("if an empty array is provided, it should never iterate, returning an empty array.", () => {
+      const result = mapWithFeedback([], (acc) => acc, "value");
+      expect(result).toEqual([]);
+    });
+
+    it("should be compatible with transformer functions", () => {
+      expect(() => {
+        const result = mapWithFeedback([1, 2, 3, 4, 5], add, 100);
+        expect(result).toEqual([101, 103, 106, 110, 115]);
+      }).not.toThrowError("Wrong number of arguments");
+    });
+
+    it("indexed should not be compatible with transformer functions", () => {
+      expect(() => {
+        const result = mapWithFeedback.indexed([1, 2, 3, 4, 5], add, 100);
+        expect(result).toEqual([101, 103, 106, 110, 115]);
+      }).toThrowError("Wrong number of arguments");
+    });
+
+    describe("type test", () => {
+      it("should return a mutable tuple type whose length matches input container's length, consisting of the type of the initial value", () => {
+        const result = mapWithFeedback(
+          [1, 2, 3, 4, 5] as const,
+          (acc, x) => acc + x,
+          100,
+        );
+        expectTypeOf(result).toEqualTypeOf<
+          [number, number, number, number, number]
+        >();
+      });
+
+      it("should return a tuple consisting of the initial value type even if the initial iterable contains a different type", () => {
+        const result = mapWithFeedback(
+          ["1", "2", "3", "4", "5"] as const,
+          (acc, x) => acc + parseInt(x),
+          100,
+        );
+        expectTypeOf(result).toEqualTypeOf<
+          [number, number, number, number, number]
+        >();
+      });
+
+      it("should correctly infer type with a non-literal array type", () => {
+        const data: Array<number> = [1, 2, 3, 4, 5];
+        const result = mapWithFeedback(data, (acc, x) => acc + x, 100);
+        expectTypeOf(result).toEqualTypeOf<Array<number>>();
+      });
+    });
+  });
+
+  describe("indexed", () => {
+    it("should track index and provide entire items array", () => {
+      let i = 0;
```

Comment:
consider using vitest's function mocking to test these sort of things instead of building your own "state":

https://vitest.dev/api/vi.html#mocking-functions-and-objects